File: src/frontend/src/types/ontology.ts
// ============================================================================
// ENUMS
// ============================================================================

export type CollectionType = 'glossary' | 'taxonomy' | 'ontology';

export type ScopeLevel = 'enterprise' | 'domain' | 'department' | 'team' | 'project' | 'external';

export type SourceType = 'custom' | 'imported';

export type ConceptStatus = 'draft' | 'under_review' | 'approved' | 'published' | 'certified' | 'deprecated' | 'archived';

export type PromotionType = 'promoted' | 'demoted' | 'migrated';

// ============================================================================
// KNOWLEDGE COLLECTION TYPES
// ============================================================================

export interface KnowledgeCollection {
  iri: string;
  label: string;
  description?: string;
  collection_type: CollectionType;
  scope_level: ScopeLevel;
  source_type: SourceType;
  source_url?: string;
  parent_collection_iri?: string;
  is_editable: boolean;
  status: string;
  
  // Audit fields
  created_at?: string;
  created_by?: string;
  updated_at?: string;
  updated_by?: string;
  
  // Computed fields
  concept_count: number;
  child_collections: KnowledgeCollection[];
}

export interface KnowledgeCollectionCreate {
  label: string;
  description?: string;
  collection_type?: CollectionType;
  scope_level?: ScopeLevel;
  parent_collection_iri?: string;
  is_editable?: boolean;
}

export interface KnowledgeCollectionUpdate {
  label?: string;
  description?: string;
  scope_level?: ScopeLevel;
  parent_collection_iri?: string;
  is_editable?: boolean;
}

// ============================================================================
// ONTOLOGY PROPERTY AND CONCEPT TYPES
// ============================================================================

export interface OntologyProperty {
  iri: string;
  label?: string;
  comment?: string;
  domain?: string;
  range?: string;
  property_type: 'datatype' | 'object' | 'annotation';
}

export interface OntologyConcept {
  iri: string;
  label?: string;  // Primary label (computed from labels dict)
  labels?: Record<string, string>;  // Multi-language labels: {"en": "Dataset", "ja": "データセット"}
  comment?: string;  // Primary comment (computed from comments dict)
  comments?: Record<string, string>;  // Multi-language comments: {"en": "A curated...", "ja": "..."}
  concept_type: 'class' | 'concept' | 'individual' | 'property' | 'term';
  source_context?: string;
  parent_concepts: string[];
  child_concepts: string[];
  related_concepts?: string[];
  properties: OntologyProperty[];
  tagged_assets: Array<{
    id: string;
    name: string;
    type?: string;
    path?: string;
  }>;
  synonyms: string[];
  examples: string[];
  
  // Property-specific fields (only set when concept_type === 'property')
  property_type?: 'datatype' | 'object' | 'annotation';
  domain?: string;
  range?: string;
  
  // Governance fields (for custom concepts)
  status?: ConceptStatus;
  version?: string;
  
  // Lifecycle timestamps
  created_at?: string;
  created_by?: string;
  updated_at?: string;
  updated_by?: string;
  published_at?: string;
  published_by?: string;
  certified_at?: string;
  certified_by?: string;
  certification_expires_at?: string;
  
  // Provenance (for promoted/migrated concepts)
  source_concept_iri?: string;
  source_collection_iri?: string;
  promotion_type?: PromotionType;
  
  // Review integration
  review_request_id?: string;
}

export interface ConceptCreate {
  collection_iri: string;
  label: string;
  definition?: string;
  concept_type?: 'class' | 'concept' | 'property' | 'individual' | 'term';
  // Property-specific fields
  property_type?: 'datatype' | 'object' | 'annotation';
  domain?: string;
  range?: string;
  synonyms?: string[];
  examples?: string[];
  broader_iris?: string[];
  narrower_iris?: string[];
  related_iris?: string[];
}

export interface ConceptUpdate {
  label?: string;
  definition?: string;
  concept_type?: 'class' | 'concept' | 'property' | 'individual' | 'term';
  // Property-specific fields
  property_type?: 'datatype' | 'object' | 'annotation';
  domain?: string;
  range?: string;
  synonyms?: string[];
  examples?: string[];
  broader_iris?: string[];
  narrower_iris?: string[];
  related_iris?: string[];
}

export interface SemanticModel {
  id: string;
  name: string;
  format: 'rdfs' | 'skos';
  original_filename?: string;
  content_type?: string;
  size_bytes?: number;
  enabled: boolean;
  created_by?: string;
  updated_by?: string;
  created_at?: string;
  updated_at?: string;
}

export interface ConceptHierarchy {
  concept: OntologyConcept;
  ancestors: OntologyConcept[];
  descendants: OntologyConcept[];
  siblings: OntologyConcept[];
}

export interface ConceptSearchResult {
  concept: OntologyConcept;
  relevance_score: number;
  match_type: 'label' | 'comment' | 'iri';
}

export interface TaxonomyStats {
  total_concepts: number;
  total_properties: number;
  taxonomies: SemanticModel[];
  concepts_by_type: Record<string, number>;
  top_level_concepts: number;
}

// Tree node structure for the UI
export interface ConceptTreeNode {
  concept: OntologyConcept;
  children: ConceptTreeNode[];
  isExpanded: boolean;
  level: number;
}

// Grouped concepts for tree view
export interface GroupedConcepts {
  [taxonomyName: string]: OntologyConcept[];
}

File: src/backend/src/models/ontology.py
from typing import Dict, List, Optional, Any
from pydantic import BaseModel, Field
from datetime import datetime
from enum import Enum


# ============================================================================
# ENUMS FOR KNOWLEDGE COLLECTIONS
# ============================================================================

class CollectionType(str, Enum):
    GLOSSARY = "glossary"
    TAXONOMY = "taxonomy"
    ONTOLOGY = "ontology"


class ScopeLevel(str, Enum):
    ENTERPRISE = "enterprise"
    DOMAIN = "domain"
    DEPARTMENT = "department"
    TEAM = "team"
    PROJECT = "project"
    EXTERNAL = "external"


class SourceType(str, Enum):
    CUSTOM = "custom"
    IMPORTED = "imported"


class ConceptStatus(str, Enum):
    DRAFT = "draft"
    UNDER_REVIEW = "under_review"
    APPROVED = "approved"
    PUBLISHED = "published"
    CERTIFIED = "certified"
    DEPRECATED = "deprecated"
    ARCHIVED = "archived"


class PromotionType(str, Enum):
    PROMOTED = "promoted"
    DEMOTED = "demoted"
    MIGRATED = "migrated"


# ============================================================================
# KNOWLEDGE COLLECTION MODELS
# ============================================================================

class KnowledgeCollection(BaseModel):
    """A glossary, taxonomy, or ontology containing concepts"""
    iri: str
    label: str
    description: Optional[str] = None
    collection_type: CollectionType = CollectionType.GLOSSARY
    scope_level: ScopeLevel = ScopeLevel.ENTERPRISE
    source_type: SourceType = SourceType.CUSTOM
    source_url: Optional[str] = None
    parent_collection_iri: Optional[str] = None
    is_editable: bool = True
    status: str = "active"
    
    # Audit fields
    created_at: Optional[datetime] = None
    created_by: Optional[str] = None
    updated_at: Optional[datetime] = None
    updated_by: Optional[str] = None
    
    # Computed fields
    concept_count: int = 0
    child_collections: List["KnowledgeCollection"] = []


class KnowledgeCollectionCreate(BaseModel):
    """Create request for a knowledge collection"""
    label: str
    description: Optional[str] = None
    collection_type: CollectionType = CollectionType.GLOSSARY
    scope_level: ScopeLevel = ScopeLevel.ENTERPRISE
    parent_collection_iri: Optional[str] = None
    is_editable: bool = True


class KnowledgeCollectionUpdate(BaseModel):
    """Update request for a knowledge collection"""
    label: Optional[str] = None
    description: Optional[str] = None
    scope_level: Optional[ScopeLevel] = None
    parent_collection_iri: Optional[str] = None
    is_editable: Optional[bool] = None


# ============================================================================
# ONTOLOGY PROPERTY AND CONCEPT MODELS
# ============================================================================

class OntologyProperty(BaseModel):
    iri: str
    label: Optional[str] = None
    comment: Optional[str] = None
    domain: Optional[str] = None
    range: Optional[str] = None
    property_type: str  # 'datatype' | 'object' | 'annotation'


class OntologyConcept(BaseModel):
    iri: str
    label: Optional[str] = None  # Primary label (computed from labels dict)
    labels: Dict[str, str] = {}  # Multi-language labels: {"en": "Dataset", "ja": "データセット"}
    comment: Optional[str] = None  # Primary comment (computed from comments dict)
    comments: Dict[str, str] = {}  # Multi-language comments: {"en": "A curated...", "ja": "..."}
    concept_type: str  # 'class' | 'concept' | 'individual' | 'term'
    source_context: Optional[str] = None  # The taxonomy/ontology source (collection IRI)
    parent_concepts: List[str] = []  # Parent class/concept IRIs
    child_concepts: List[str] = []   # Child class/concept IRIs
    related_concepts: List[str] = [] # skos:related IRIs
    domain: Optional[str] = None     # For properties: rdfs:domain
    range: Optional[str] = None      # For properties: rdfs:range
    properties: List[OntologyProperty] = []
    tagged_assets: List[Dict[str, Any]] = []  # Linked data assets
    synonyms: List[str] = []
    examples: List[str] = []
    
    # Governance fields (for custom concepts)
    status: Optional[ConceptStatus] = None
    version: Optional[str] = None
    
    # Lifecycle timestamps
    created_at: Optional[datetime] = None
    created_by: Optional[str] = None
    updated_at: Optional[datetime] = None
    updated_by: Optional[str] = None
    published_at: Optional[datetime] = None
    published_by: Optional[str] = None
    certified_at: Optional[datetime] = None
    certified_by: Optional[str] = None
    certification_expires_at: Optional[datetime] = None
    
    # Provenance (for promoted/migrated concepts)
    source_concept_iri: Optional[str] = None
    source_collection_iri: Optional[str] = None
    promotion_type: Optional[PromotionType] = None
    
    # Review integration
    review_request_id: Optional[str] = None


class ConceptCreate(BaseModel):
    """Create request for a concept"""
    collection_iri: str
    label: str
    definition: Optional[str] = None
    concept_type: str = "concept"  # 'class' | 'concept' | 'term'
    synonyms: List[str] = []
    examples: List[str] = []
    broader_iris: List[str] = []  # skos:broader
    narrower_iris: List[str] = []  # skos:narrower
    related_iris: List[str] = []  # skos:related


class ConceptUpdate(BaseModel):
    """Update request for a concept"""
    label: Optional[str] = None
    definition: Optional[str] = None
    synonyms: Optional[List[str]] = None
    examples: Optional[List[str]] = None
    broader_iris: Optional[List[str]] = None
    narrower_iris: Optional[List[str]] = None
    related_iris: Optional[List[str]] = None
    
    
class SemanticModel(BaseModel):
    """Represents a loaded semantic model (taxonomy/ontology source)"""
    name: str
    description: Optional[str] = None
    source_type: str  # 'file' | 'database' | 'external' | 'schema'
    format: Optional[str] = None  # 'ttl' | 'rdf' | 'owl'
    concepts_count: int = 0
    properties_count: int = 0
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


class ConceptHierarchy(BaseModel):
    """Represents hierarchical relationships for visualization"""
    concept: OntologyConcept
    ancestors: List[OntologyConcept] = []
    descendants: List[OntologyConcept] = []
    siblings: List[OntologyConcept] = []


class ConceptSearchResult(BaseModel):
    concept: OntologyConcept
    relevance_score: float = 0.0
    match_type: str = "label"  # 'label' | 'comment' | 'iri'


class TaxonomyStats(BaseModel):
    total_concepts: int
    total_properties: int
    taxonomies: List[SemanticModel]
    concepts_by_type: Dict[str, int] = {}
    top_level_concepts: int = 0